lib/jarvis/commands/publish.rb

```diff
@@ -29,28 +42,139 @@ def execute
+
+      unless force?
+        build = build_report(project)
+        workdir_sha1 = Jarvis::Git.sha1(workdir) 
+
+        if build.sha1 == workdir_sha1
+          raise CommitHashDontMatch, "workdir_sha1: #{workdir_sha1}, build_sha1: #{build_sha1}"
+        end
+
+        if !build.success?
+          raise CIBuildFail, "Expecting success got #{build.status}"
+        end
+      end
+
       logger.info("Switching branches", :branch => branch)
       git.checkout(branch)
       context = logger.context
       context[:operation] = "publish"
       context[:branch] = branch
 
-      TASKS.each do |command|
-        context[:command] = command
-        Jarvis.execute(command, logger, git.dir)
+      TASKS.each do |command, condition|
+        if condition.call(workdir)
+          context[:command] = command
+          Jarvis.execute(command, logger, git.dir)
 
-        # Clear the logs if it was successful
-        logs.clear unless logger.debug?
+          # Clear the logs if it was successful
+          logs.clear unless logger.debug?
+        end
       end
       context.clear()
       git.reset
       git.clean(force: true)
+
+      check_version_match
     end
 
-    puts I18n.t("lita.handlers.jarvis.publish success", organization: project.organization, project: project.name, branches: branches.join(", "))
+
+    puts I18n.t("lita.handlers.jarvis.publish success",
+                :organization => project.organization,
+                :project => project.name,
+                :branches => branches.join(", "))
   rescue => e
-    puts I18n.t("lita.handlers.jarvis.exception", :exception => e.class, :message => e.to_s, :stacktrace => e.backtrace.join("\n"), :command => "merge", :logs => logs.collect { |l| l[:message] }.join("\n"))
+    puts I18n.t("lita.handlers.jarvis.exception", :exception => e.class,
+                :message => e.to_s,
+                :stacktrace => e.backtrace.join("\n"),
+                :command => "publish",
+                :logs => logs.collect { |l| l[:message] }.join("\n"))
+  end
+
+  def contains_vendor_files?(directory)
+  end
+
+  def check_version_match
+    name, local_version = gem_specification
+    published_gems = Gems.versions(name)
+    # First version out
+    return if published_gems == NO_PREVIOUS_GEMS_PUBLISHED 
+    remote_versions = published_gems.collect { |v| v["number"] }
+
+    if !remote_versions.include?(local_version)
+      raise RemoteVersionDontMatchLocal, "local version: #{local_version}, not in remove version #{remote_versions}"
+    end
+  end
+
+  def gem_specification
+    # HACK: if you are using the `real` bundler way of creating gem
+    # You have to create a version.rb file containing the version number
+    # and require the file in the gemspec. 
+    # Ruby will cache this require and not reload it again in a long running
+    # process like the bot.
+    cmd = "ruby -e \"spec = Gem::Specification.load('#{gemspec}'); puts [spec.name, spec.version].join(',')\""
```

Comment:
I forget; is there a reason why we shell out to ruby here? Because we can't load some gemspecs under MRI (JRuby-only?)?